org.osgi.test.assertj/src/main/java/org/osgi/test/assertj/bundle/AbstractBundleAssert.java

```
@@ -0,0 +1,366 @@
+	}
+
+	public SELF isNotInState(int expected) {
+		isNotNull();
+		final String expectedState = stateToString(expected);
+		if (actual.getState() == expected) {
+			failWithMessage("%nExpecting%n  <%s>%nnot to be in state:%n  <%s>%n but it was", actual,
+				expectedState);
+		}
+		return myself;
+	}
+
+	public SELF isInStateMask(int mask) {
+		isNotNull();
+		if (mask <= 0 || mask >= ACTIVE * 2) {
+			throw new IllegalArgumentException("Mask testing for an illegal state: " + mask);
+		}
+		if ((actual.getState() & mask) == 0) {
+			final String states = IntStream.of(STATES)
+				.filter(x -> (x & mask) != 0)
+				.mapToObj(AbstractBundleAssert::stateToString)
+				.collect(joining(","));
+			failWithMessage("%nExpecting%n  <%s>%nto be in one of states:%n  [%s]%n but was in state:%n  <%s>",
+				actual,
+				states, stateToString(actual.getState()));
+		}
+		return myself;
+	}
+
+	public SELF isNotInStateMask(int mask) {
+		isNotNull();
+		if (mask <= 0 || mask >= ACTIVE * 2) {
+			throw new IllegalArgumentException("Mask testing for an illegal state: " + mask);
+		}
+		if ((actual.getState() & mask) != 0) {
+			final String states = IntStream.of(STATES)
+				.filter(x -> (x & mask) != 0)
+				.mapToObj(AbstractBundleAssert::stateToString)
+				.collect(joining(","));
+			failWithMessage("%nExpecting%n  <%s>%nto not be in one of states:%n  [%s]%n but was in state:%n  <%s>",
+				actual,
+				states, stateToString(actual.getState()));
+		}
+		return myself;
+	}
+
+	private static final int[] STATES = {
+		UNINSTALLED, INSTALLED, RESOLVED, STARTING, STOPPING, ACTIVE
+	};
+
+	public SELF isInOneOfStates(int... states) {
+		return isInStateMask(IntStream.of(states)
+			.filter(x -> stateToString(x) != null) // Will throw IAE if
+													// necessary
+			.reduce(0, (x, y) -> x | y));
+	}
+
+	public SELF isNotInOneOfStates(int... states) {
+		return isNotInStateMask(IntStream.of(states)
+			.filter(x -> stateToString(x) != null) // Will throw IAE if
```

If you need to see of the supplied state value is valid, just AND it with the OR of all the valid state values. Must faster and simpler.

---

Again, the idea was that each of the states would be a single bit and that if they weren't it would throw IAE (the call to `stateToString()` was a semi-hack to trigger the IAE). If you wanted to allow test masks, you can use `isInStateMask()`.

---

Still, there is no need for strings or switches here. As I mention above, you can test that each state arg has exactly one bit set and error if not.

---

The main idea of the string (and the switch that generates it) is to produce a human-friendly error message in the event of a failing assertion (naming the state(s) instead of printing a number).

I'll try and simplify it.

---

Done.